### PR TITLE
TP-462: Allow the importer to run in a distributed system.

### PR DESCRIPTION
**Pull request by NDevox**
To enable the importer to run in a distributed manner it needs
access to the chunks of XML to import through a shared system.
The tasks to run the import must also be shared in a similar way.

This commit adds importer database models which hold chunks of XML
that all instances of the system can access. These chunks are split
firstly on size (currently 50MB is the default). However if the file
being imported can be done so in an asynchronous manner it is then
also split on record code and (for measures and commodities) chapter
heading. These chunks can then be imported simultaneously
(asynchronous imports are almost solely intended for seed files).

The function for running an import has been converted into a Celery
task so this can also be handled in a distributed manner.

The importer caching system, for caching individual models, has been
moved to the nursery - so as to de-pollute the models system.
Some business rules and handlers have been further updated.

### CODE REVIEW

File: settings/common.py [hunk 1]
```diff
@@ -330,3 +330,4 @@
 }
 
 SKIP_VALIDATION = is_truthy(os.getenv("SKIP_VALIDATION", False))
+SKIP_WORKBASKET_VALIDATION = is_truthy(os.getenv("SKIP_WORKBASKET_VALIDATION", True))
```
Comment (simonwo) [inline settings/common.py:333]: Is `True` the right default?

**Review by simonwo — APPROVED** [review]
Looks good from my breeze through. I know that @stuaxo has also done a review and you've updated based on that?

File: importer/management/commands/chunk_taric.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    )
+
+    chunk.write(
+        etree.tostring(
+            transaction
+        )  # pythons XML doesn't write namespaces back correctly.
+        .replace(b"<ns0:", b"<env:")
+        .replace(b"<ns1:", b"<ns2:")
+        .replace(b"</ns0:", b"</env:")
+        .replace(b"</ns1:", b"</ns2:")
```
[the remark this answers is not in the document]
Comment (NDevox) [reply importer/management/commands/chunk_taric.py:231]: Yeah I was pretty upset by this. We essentially already use two XML libraries for parsing and writing.

The default python lib (`import xml`) is slower, and doesn't store namespaces properly - so we have to do this.

The lxml lib (`import lxml`) is faster and keeps the namespaces, but doesn't handle some of the encoding well (we have some badly encoded HTML in our data) and ends up breaking the file when writing it out.

File: importer/management/commands/chunk_taric.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        parser.add_argument(
+            "-s",
+            "--split-codes",
+            help="Split the file based on record codes",
+            action="store_true",
+        )
+
+    def handle(self, *args, **options):
+        with open(options["taric3_file"], "rb") as taric3_file:
+            chunk_taric(taric3_file=taric3_file, split_on_code=options["split_codes"])
```
[the remark this answers is not in the document]
Comment (NDevox) [reply importer/management/commands/chunk_taric.py:297]: That's actually what it was doing before :-D If you go back one commit there is a `split_taric` command which does exactly that. Could probably implement an option to do it for us, won't be much work (another PR though)

File: importer/tasks.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    status: str,
+    username: str,
+):
+    """
+    Task for importing an XML chunk into the database.
+
+    This task must ensure the chunks status reflects the import process,
+    whether it is currently running, errored or done. Once complete it is
+    also responsible for finding and setting up the next chunk tasks.
+    """
```
[the remark this answers is not in the document]
Comment (NDevox) [reply importer/tasks.py:26]: No, nothing stopping us though - they're not protected. Just think it's not worthwhile.

File: importer/utils.py [hunk 5]
```diff
@@ -57,3 +60,38 @@ def generate_key(
...
+        if subclass.record_code not in dependency_map:
+            dependency_map[subclass.record_code] = set()
+        for _, relation in subclass.get_relations():
+            if (
+                relation.record_code != subclass.record_code
+                and relation.record_code in record_codes
+            ):
+                dependency_map[subclass.record_code].add(relation.record_code)
+
+    return dependency_map
```
[the remark this answers is not in the document]
Comment (NDevox) [reply importer/utils.py:94]: I can see it being useful for a lot of things.

Also started implementing a deletion system which can figure out dependencies and delete models in the right order... but that's only really useful for debugging in dev.

File: settings/common.py (hunk 1 above)
Comment (NDevox) [reply settings/common.py:333]: Probably not, I'll change that.

File: importer/management/commands/chunk_taric.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    try:
+        chunk = chunks_in_progress[key]
+    except KeyError:
+        chunk = TemporaryFile()
+        chunk.write(
+            f"""\
+<?xml version='1.0' encoding='UTF-8'?>
+<env:envelope xmlns:env="urn:publicid:-:DGTAXUD:GENERAL:ENVELOPE:1.0" \
+xmlns:ns2="urn:publicid:-:DGTAXUD:TARIC:MESSAGE:1.0" id="{batch.name}">
+""".encode()
```
[the remark this answers is not in the document]
Comment (NDevox) [reply importer/management/commands/chunk_taric.py:45]: I have doubts but will take a look.

**Pull request merged**